tests/unit/test_charm.py

```diff
@@ -0,0 +1,145 @@
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    # Event handlers
+    @patch("charm.TrainingOperatorCharm._create_crds")
+    @patch("charm.TrainingOperatorCharm._create_resource")
+    def test_install_event(self, create_auth, create_crd):
+        self.harness.charm.on.install.emit()
+
+        # Ensure create_auth and create_crds
+        # are called once
+        create_auth.assert_called_once()
+        create_crd.assert_called_once()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating auth resources
+        create_auth.side_effect = _FakeApiError()
+        self.harness.charm.on.install.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Creating/patching resources failed with code {create_auth.side_effect.response.code}."))
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating CRDs
+        create_crd.side_effect = _FakeApiError()
+        self.harness.charm.on.install.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Creating/patching resources failed with code {create_crd.side_effect.response.code}."))
+
+    @patch("charm.TrainingOperatorCharm._update_layer")
+    @patch("charm.TrainingOperatorCharm._patch_resource")
+    def test_config_changed_event(self, patch, update):
+        self.harness.charm.on.config_changed.emit()
+        update.assert_called_once()
+
+        # Run each case of patch here?
+        patch.assert_called()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when patching auth resources
+        patch.side_effect = _FakeApiError()
+        self.harness.charm.on.config_changed.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Patching resources failed with code {patch.side_effect.response.code}."))
+
+    @patch("charm.TrainingOperatorCharm._update_layer")
+    def test_on_training_operator_pebble_ready(self, update):
+        self.harness.container_pebble_ready("training-operator")
+
+        # Is this assertion enough? All the method
+        # does is update the layer and restart the service
+        self.assertIsNotNone(self.harness.get_container_pebble_plan("training-operator")._services)
+
+    # Helpers
+    @patch("charm.Client.create")
+    def test_create_resource(self, client: MagicMock):
+        self.harness.charm._context = {"namespace": "kubeflow", "app_name": "unit-training-operator"}
+        self.harness.charm._create_resource(resource_type="auth", context=self.harness.charm._context)
+        self.harness.charm._create_resource(resource_type="crds")
+
+        # Ensure all resources listed in src/*.yaml exist
+        resources = []
+        for manifest in glob("src/*.yaml"):
+            with open(manifest) as f:
+                resources.extend([r for r in codecs.load_all_yaml(f, self.harness.charm._context)])
+        for resource in resources:
```

Comment:
I don't know if there's an equivalent in UnitTest, but this seems like where I'd use a parametrized test in pytest, testing only a single resource call at a time.  Something like:

(syntax is probably wrong, just going by memory)
```python
@patch("charm.Client.create")
@pytest.mark.parametrize(
  "resource,context,yaml_files",
  [
    ("auth", {"namespace": "kubeflow, ...}, [yaml_file1, ...]),
    ("crds", {}, [yaml_file1, ...]),
  ]
)
def test_create_resource(resource, context, yaml_files, client: MagicMock):
  self.harness.charm._create_resource(resource_type=resource, context=context)
  resources = []
  for manifest in yaml_files:
  ...
```

(and then do the failure tests separately, again parametrized if you want to test both types)

Reply:
We can so something similar in unittest with [subTest()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest). I will explore this option, I like it better.

Reply:
this is correct syntax.  I prefer this too and use it quite frequently